include/boost/math/differentiation/autodiff.hpp

```diff
@@ -1595,35 +1642,63 @@ class numeric_limits<boost::math::differentiation::detail::fvar<RealType,Order>>
     : public numeric_limits<typename boost::math::differentiation::detail::fvar<RealType,Order>::root_type>
 { };
 
-} // namespace std
+}  // namespace std
 
-namespace boost { namespace math { namespace tools {
+namespace boost {
+namespace math {
+namespace tools {
 
 // See boost/math/tools/promotion.hpp
 template <typename RealType0, size_t Order0, typename RealType1, size_t Order1>
-struct promote_args_2<differentiation::detail::fvar<RealType0,Order0>,differentiation::detail::fvar<RealType1,Order1>>
-{
-    using type = differentiation::detail::fvar<typename promote_args_2<RealType0,RealType1>::type,
+struct promote_args_2<differentiation::detail::fvar<RealType0, Order0>,
+                      differentiation::detail::fvar<RealType1, Order1>> {
+  using type = differentiation::detail::fvar<typename promote_args_2<RealType0, RealType1>::type,
 #ifndef BOOST_NO_CXX14_CONSTEXPR
-        std::max(Order0,Order1)>;
+                                             std::max(Order0, Order1)>;
 #else
-        Order0 < Order1 ? Order1 : Order0>;
+        Order0<Order1 ? Order1 : Order0>;
 #endif
 };
 
 template <typename RealType0, size_t Order0, typename RealType1>
-struct promote_args_2<differentiation::detail::fvar<RealType0,Order0>,RealType1>
-{
-    using type = differentiation::detail::fvar<typename promote_args_2<RealType0,RealType1>::type,Order0>;
+struct promote_args_2<differentiation::detail::fvar<RealType0, Order0>, RealType1> {
+  using type = differentiation::detail::fvar<typename promote_args_2<RealType0, RealType1>::type, Order0>;
 };
 
 template <typename RealType0, typename RealType1, size_t Order1>
-struct promote_args_2<RealType0,differentiation::detail::fvar<RealType1,Order1>>
-{
-    using type = differentiation::detail::fvar<typename promote_args_2<RealType0,RealType1>::type,Order1>;
+struct promote_args_2<RealType0, differentiation::detail::fvar<RealType1, Order1>> {
+  using type = differentiation::detail::fvar<typename promote_args_2<RealType0, RealType1>::type, Order1>;
 };
 
-} } } // namespace boost::math::tools
+template <typename RealType, std::size_t Order>
+struct promote_args<differentiation::detail::fvar<RealType, Order>> {
+  using type = differentiation::detail::fvar<typename promote_args<RealType>::type, Order>;
+};
+
+template <typename ToType, typename RealType, std::size_t Order>
+inline ToType real_cast(const differentiation::detail::fvar<RealType, Order>& from_v) {
+  return static_cast<ToType>(static_cast<RealType>(from_v));
+}
+
+}  // namespace tools
+
+namespace policies {
```

Comment:
I hope the `boost::math` functions are updated to not require Policy objects.

Reply:
This pattern isn't uncommon, but can be demonstrated here (hermite.hpp):
   `typedef typename tools::promote_args<T>::type result_type;`
   `typedef typename policies::evaluation<result_type, Policy>::type value_type;`

If `T` is a type alias for `float`, `value_type` is a type alias for `double`. Without the policy specializations for `autodiff_fvar<float,Order>`, `value_type` is an alias for `autodiff_fvar<float,Order>`. The difference doesn't become obvious unless we check (as we do) that if `T` causes a function to throw an exception, we should expect its `autodiff_fvar` analog to cause the same behavior (in many cases, this wasn't the case; the `autodiff_fvar` would just remain intact with its array filled with non-finite values -- so nonsense with neither floating point errors or exceptions).

Reply:
Ok thanks for the explanation.